dbschema/project.gel

```diff
@@ -121,6 +121,149 @@ module default {
         projectContext := __new__.projectContext,
       }
     );
+
+    trigger createPeriodicReports after insert for each do (
+      with
+        interval := (select 
+          if __new__.financialReportPeriod = default::ReportPeriod.Monthly then '1' else '3'),
+        reportRanges := Project::create_periodic_report_ranges(
+          __new__.mouStart,
+          __new__.mouEnd,
+          interval
+        )
+      for reportRange in reportRanges 
+      union (
+        (insert default::FinancialReport {
+          createdAt := datetime_of_statement(),
+          modifiedAt := datetime_of_statement(),
+          createdBy := assert_exists(global currentActor),
+          modifiedBy := assert_exists(global currentActor),
+          project := __new__,
+          projectContext := __new__.projectContext,
+          container := __new__,
+          period := reportRange
+        }),
+        (insert default::NarrativeReport {
+          createdAt := datetime_of_statement(),
+          modifiedAt := datetime_of_statement(),
+          createdBy := assert_exists(global currentActor),
+          modifiedBy := assert_exists(global currentActor),
+          project := __new__,
+          projectContext := __new__.projectContext,
+          container := __new__,
+          period := reportRange
+        })
+      )
+    );
+
+    trigger addRemovePeriodicReports after update for each 
+      when (
+        __old__.mouStart ?!= __new__.mouStart
+        or __old__.mouEnd ?!= __new__.mouEnd
+        or __old__.financialReportPeriod ?!= __new__.financialReportPeriod
+      ) 
+      do (
+      with
+        existingReports := (
+            select PeriodicReport
+            filter .container.id = __old__.id
+        ),
+        interval := (
+          select (if __new__.financialReportPeriod = default::ReportPeriod.Monthly then '1' else '3')
+          ),
+        requestedReportPeriods := Project::create_periodic_report_ranges(
+          __new__.mouStart,
+          __new__.mouEnd,
+          interval
+        ),
+      result := (if __old__.financialReportPeriod ?!= __new__.financialReportPeriod then (
+        with
+          reportPeriodsWithoutFiles := (
+            select existingReports
+            filter not exists .reportFile
+          ),
+          deletedReportPeriods := (
+            for reportPeriod in reportPeriodsWithoutFiles 
+            union (
+              delete reportPeriod
+            )
+          ),
+          financialReports := (for reportPeriod in requestedReportPeriods
```

Comment:
@CarsonF - I don't like that we're looping through the financials and the narratives separately here.  If there's a way to combine them both into one loop please let me know.  I wanted to do that (if it is possible) before I abstract this functionality out into it's own function.

Reply:
I do not think it's possible to unify the insert clauses, or dynamically reference different types for inserts.

I _think_ the loops could be combined though if that's what you mean.
```edgeql
new := (
  for reportPeriod in requestedReportPeriods union ({
    insert Fin {},
    insert Narr {},
  })
)
```

Reply:
Yeah, that's the structure I originally had, but was running into this kind of error

> operator 'std::IF' cannot be applied to operands of type 'tuple<default::FinancialReport, default::NarrativeReport>', 'std::bool', 'default::PeriodicReport'

on account of Gel not being able to reconcile that the else statement could only return in reality FR & NR.

The solution was to split them out and manually merge them like this

> select financialReports union narrativeReports

Reply:
Gotcha. Then that's probably necessary.